New pips, second try

msalle opened the pull request:
Hi Andrea,
hereby a new pull request.

msalle committed:
Adding two new PIPs

Adding X509ExtractorPIP and PolicyNamesPIP with corresponding
*IniConfigurationParser classes and *Test classes.
PolicyNamesPIP makes use of a sub package policynamespip for a caching the info
files.
This solves https://github.com/argus-authz/argus-pep-server/issues/15.

msalle committed:
Add new PIPs to pepd.ini

Add new PIPs including default config to pepd.ini

msalle committed:
Update license to match template

msalle committed:
Increase junit version for new style junit tests

junit framework is deprecated in favour of junit tests indicated with @test etc.
This requires a newer version of junit.

msalle committed:
Replace deprecated getValue with getObjectValue

msalle committed:
Add meta-policy files combining the standard IGTF

Add two new policy files which combine the standard IGTF policies, in order to
simplify Argus PAP policy matching: WLCG VOs can be matched using
policy-aspen-birch-cedar-dogwood while other VOs need to combine with
policy-aspen-birch-cedar.

msalle committed:
Need to check whether info file exists

Before trying to get the modification time, test whether the file actually
exists.

msalle committed:
Split-off cache update from PIP

Introduce intermediate class UpdatingPolicyNamesCache that takes care of the
updating of the PolicyNamesCache (formerly PolicyNamesPIPCache).
The PIP itself now does not need to keep track of that. Remove the unused and
non-trivial possibility of updating the trust dir for a running PIP.

msalle committed:
Improve locking using ReentrantReadWriteLock

Implement proper read/write locks to prevent memory corruption.

andreaceccanti commented on src/main/assembly/fakeroot.xml:
```diff
@@ -103,6 +103,12 @@ Installed in a fakeroot directory structure.
+	<!-- .info files -->
+        <fileSet>
+            <directory>src/main/infofiles</directory>
+            <outputDirectory>/etc/grid-security/certificates</outputDirectory>
```
I don't like very much the idea that the info files end up in /etc/grid-security/certificates, and that
Argus owns stuff in that directory.

Are these files argus specific? If so, they could live in /etc/argus. If not, they shouldn't be packaged in Argus.

msalle replied:
they are indeed Argus specific, in that they don't (and will not) come from IGTF, but they do have the status of policy .info files in the same way as the policy-igtf-classic.info. Having them in /etc/argus will not work therefore, since they would not be found. The  reason for having them in the first place, is to prevent us from needing separate permit lines for each individual policy (e.g. classic, mics, slcs and iota) but instead making a single entry with e.g. policy-aspen-birch-cedar-dogwood.
What we could do is put the file itself in /etc/argus/... and make a symlink into the /etc/grid-security/certificates directory.

andreaceccanti replied:
Since these files are fully argus-specific, then it's up to the PIP to look them up in whatever directory we choose, or am I missing something? I would go for something into /etc/argus/, but also /etc/grid-security/argus/ would be fine for me (following an approach similar to /etc/grid-security/vomsdir).

msalle replied:
I've discussed with DavidG and we could try to push them via the EGI policy packages (such as ca-policy-egi-core-1.79-1.noarch.rpm), which are specific to EGI, and actually also produced by David with an EGI hat instead of IGTF hat. The point is that the PIP looks in the 'trustDir' (i.e. normally /etc/grid-security/certificates) for all the info files, and hence that's also where these files should end up, either as file or as symlink. Any other directory does not make sense since they are the same sort of file from the PIP perspective. They are (reasonably) normal info files, only not IGTF.

andreaceccanti replied:
> The point is that the PIP looks in the 'trustDir' (i.e. normally /etc/grid-security/certificates) for all the info file

The point is: are these files meant to be used by other services? If not, they should be packaged in an argus-owned directory, and the PIP modified to look into this directory.
I think the fact that the policy files and certificates must live in the same dir is only a limitation of the current PIP policy lookup implementation, or there's more to it?

msalle replied:
First of all, they might be used by other services, in that sense they're not specific to Argus, hence the suggestion to ship them with the EGI policy packages. For most services they are probably not needed, Argus needs them due to the limitation that we can _not_ make logical OR statements in a permit rule, something along the lines of
```
rule permit {
      vo="atlas"
      { ca-policy-names="policy-igtf-classic" OR
        ca-policy-names="policy-igtf-mics" OR
        ca-policy-names="policy-igtf-slcs" OR
        ca-policy-names="policy-igtf-iota"
      }
  }
```
Furthermore, info files have always been in the same directory as the pem files, that includes policy info files such as policy-igtf-classic.info but also the info files for the CAs themselves.

andreaceccanti replied:
Ok, so it seems we have an agreement that these files should be packaged outside of the Argus PEPD, in a different package, provided by EGI, so please remove them from the assembly file.

msalle replied:
done. Same thing for the src/main/assembly/standalone.xml

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	throws PIPProcessingException
+    {
+	long t0=System.nanoTime();
+	boolean pipprocessed=false;
```
Java favours camelCase, so this should be pipProcessed.

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+    {
+	long t0=System.nanoTime();
+	boolean pipprocessed=false;
+	String issuerdn=null;
```
issuerdn -> issuerDn

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	    for (Attribute attr: attributes) {
+		if (ATTR_X509_ISSUER.equals(attr.getId())) {
+		    // Take first value (it should be singlevalued)
+		    Object tmp = attr.getValues().iterator().next();
```
tmp is a bad name, probably  x509IssuerAttr is a better alternative

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	    }
+
+	    // Look for the issuerdn in the .info files
+	    String[] policynames=new String[0];
```
policynames -> policyNames

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	    }
+
+	    // Create new attribute and add the policy names
+	    Attribute attr_policynames =
```
attr_policynames -> policyNamesAttr

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	    // Add to the current subject
+	    attributes.add(attr_policynames);
+	    pipprocessed=true;
+	    log.debug("Added attribute \""+attribute_name+"\"");
```
Use slf4j placeholders for the logging statements:
```java
log.debug("Added attribute \"{}\"", attributeName)
```

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	}
+
+	// Log statistics
+	log.debug("PIP parsing took "+(System.nanoTime()-t0)/1000000.0+" msec");
```
see comment above

msalle replied:
not updated, see also above

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+	}
+
+	// Log statistics
+	log.debug("PIP parsing took "+(System.nanoTime()-t0)/1000000.0+" msec");
```
Use TimeUnit to do time conversion

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/TimeUnit.html

msalle replied:
That would be reasonable, but rounds down (truncates) to a whole msec. Hence I prefer to do the division.

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+    implements IniSectionConfigurationParser<PolicyInformationPoint>
+{
+    /** Class logger. */
+    private Logger log = LoggerFactory.getLogger( PolicyNamesPIPIniConfigurationParser.class);
```
Logger should be static and final. 
Static and final variables are typically named all in uppercase, so this becomes:
```java 
private static final Logger LOG =  LoggerFactory.getLogger( PolicyNamesPIPIniConfigurationParser.class);
```

msalle replied:
see above, we need to decide what to do with the other PIPs.

andreaceccanti replied:
I would have the logger static and final for this PIP, since we don't plan to have multiple instances in a PIP chain, do we, and even in that case, do we want to have logging specific to the instance? If so, then the log calls should be changed to include the PIP instance name , probably...

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIP.java:
```diff
@@ -0,0 +1,242 @@
+ */
+public class PolicyNamesPIP extends AbstractPolicyInformationPoint {
+    /** Class logger instance */
+    private final Logger log = LoggerFactory.getLogger(PolicyNamesPIP.class);
```
logger should be static

msalle replied:
Are you sure? None of the other PIPs have it declared as such, they're either private or private final.

andreaceccanti replied:
Typically loggers are static and final, and shared among objects of the same class, but you can have also a logger specific to each instance, it depends on how you want to log. 
AFAIU we'll have only one PIP instance of the PolicyNamesPIP in a PEPD, so the logger can stay non-static, you wont' have two PIP objects and two loggers. Even in this case I usually prefer to go for static and final.

msalle replied:
I also believe there is only one PIP instance, although you probably know better. It's fine with me to make it static and final (I will), but would then strongly suggest we do the same for all the other PIPs currently in the PEPd code base.

andreaceccanti replied:
Ok!

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+    public PolicyInformationPoint parse(Ini.Section iniConfig, AbstractConfigurationBuilder<?> configBuilder)
+		    throws ConfigurationException {
+
+	    long updateinterval_long = -1;
```
updateinterval_long -> updateIntervalLong

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+
+	    // Get configuration values
+	    String id = iniConfig.getName();
+	    String trust_dir = iniConfig.get(TRUSTDIR_KEY);
```
trust_dir -> trustDir

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+	    // Get configuration values
+	    String id = iniConfig.getName();
+	    String trust_dir = iniConfig.get(TRUSTDIR_KEY);
+	    String updateinterval = iniConfig.get(UPDATEINTERVAL_KEY);
```
updateinterval -> updateInterval

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+		try {
+		    updateinterval_long = Integer.parseInt(updateinterval);
+		} catch (NumberFormatException e)	{
+		    throw new ConfigurationException(
```
Using String.format can make things more readable, like:
```java
throw new ConfigurationException(String.format("Cannot convert  %s = %s to a long", UPDATEINTERVAL_KEY, updateInterval));
```
which becomes:
```java
throw new ConfigurationException(format("Cannot convert  %s = %s to a long", UPDATEINTERVAL_KEY, updateInterval));
```
if you statically import the String.format static method

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+		    pip = new PolicyNamesPIP(id, trust_dir);
+		}
+	    } catch (IOException e) {
+		throw new ConfigurationException(
```
I would log the exception, and chain it properly before rethrowing:
```java
final String errorMsg = String.format("Could not instantiate PIP: %s", e.getMessage());
LOG.error(errorMsg, e);
throw new ConfigurationException(errorMsg, e);
```

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,132 @@
+
+	    // Set attribute name
+	    if (attributeName != null)  {
+		log.debug("Found "+ATTRIBUTENAME_KEY+" = "+attributeName);
```
Use slf4j placeholder

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+ */
+public class X509ExtractorPIP extends AbstractPolicyInformationPoint {
+    /** Class logger instance. */
+    private final Logger log = LoggerFactory.getLogger(X509ExtractorPIP.class);
```
static final Logger LOG

msalle replied:
see above

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+     */
+    public boolean populateRequest(Request request) throws PIPProcessingException {
+	long t0=System.nanoTime();
+	boolean pipprocessed=false;
```
pipprocessed -> pipProcessed

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	    log.error("Request has no subjects");
+	    throw new PIPProcessingException("No subject found in request");
+	}
+	if (subjects.size()>1)
```
braces around all if statements:
```java
if (subjects.size()>1){
...
}
```

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	    throw new PIPProcessingException("No subject found in request");
+	}
+	if (subjects.size()>1)
+	    log.warn("Request has "+subjects.size()+" subjects, taking first match");
```
use slf4j placeholders

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	
+	// Loop over all subjects to look for end-entity certificate
+	for (Subject subject : subjects) {
+	    if (subject==null)
```
braces around all block statements

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+		continue;
+	    Set<Attribute> attributes = subject.getAttributes();
+	    X509Certificate cert = getCertFromSubject(attributes);
+	    if (cert == null)
```
again, braces around all block statements

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+			}
+			Attribute attrCAPolicyOids = new Attribute(ATTR_CA_POLICY_OID);
+			Set<Object> values = attrCAPolicyOids.getValues();
+			for (int j=0; j<oids.length; j++)
```
braces around all block statements

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+			attributes.add(attrCAPolicyOids);
+			pipprocessed=true;
+			// Log that we succeeded
+			log.debug("Added attribute \""+ATTR_CA_POLICY_OID+"\" ("+oids.length+" value(s))");
```
Use slf4j placeholders

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+			attributes.add(attrIssuerDN);
+			pipprocessed=true;
+			// Log that we succeeded
+			log.debug("Added attribute \""+ATTR_X509_ISSUER+"\" ("+value+")");
```
Use slf4j placeholders

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+			log.debug("Added attribute \""+ATTR_X509_ISSUER+"\" ("+value+")");
+			break;
+		    default:
+			throw new PIPProcessingException("Unknown attribute "+acceptedAttrIDs[i]+" specified");
```
log before throwing

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	// Log statistics
+	log.debug("PIP parsing took "+(System.nanoTime()-t0)/1000000.0+" msec");
```
Use SLF4J placeholders and TimeUnit for conversions

msalle replied:
Done for placeholder, not for TimeUnit (see before).

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+     */
+    private X509Certificate getCertFromSubject(Set<Attribute> attributes)	{
+	// Protect against empty set
+	if (attributes==null)
```
braces around all block statements

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+
+	// Loop over all attributes, looking for ATTR_X509_ISSUER
+	for (Attribute attr: attributes) {
+	    if (attr==null)
```
braces around all block statements

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+		if (value.startsWith("-----BEGIN "))	{
+		    // assume WN profile: single PEM encoded chain. Should also
+		    // be single-valued according to https://edms.cern.ch/document/1058175/1.0.1
+		    if (attributeValues.size() > 1)
```
braces around all block statements

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+		} catch (Exception e) {
+		    // This might be a IOException, but also for invalid base64
+		    // a StringIndexOutOfBoundsException or a DecoderException
+		    log.error("Skipping invalid key-info attr: Parsing value as a certificate failed: "+e.getMessage());
```
Log exception as follows:
```bash
LOG.error("Skipping invalid key-info attr: Parsing value as a certificate failed", e);
```

msalle replied:
Do you really want that, it will give quite extensive stack traces from e.g. BouncyCastle? Why not just print the actual error message?

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	List<String> oidList = new LazyList<String>();
+
+	// OID for certificate_policies (=2.5.29.32)
+//	String certPolicies = X509Extension.certificatePolicies.toString();
```
Drop commented out lines (we have git!)

msalle replied:
agreed, done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	// Grab bare extension value from certificate
+	byte[] extvalue = cert.getExtensionValue(certPolicies);
+	if (extvalue==null)
+	    return null;
```
braces around block

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+
+	// Try to parse the raw bytes as ASN1Sequence
+	ASN1Sequence seq;
+	try {
```
This parsing logic should be in a separate helper method IMO

msalle replied:
That's fine with me, but which part: only obtaining the ASN1Sequence seq, or including the conversion from ASN1Sequence to list of OIDs?

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	    DEROctetString oct=(DEROctetString)(new ASN1InputStream(new ByteArrayInputStream(extvalue)).readObject());
+	    seq = (ASN1Sequence) new ASN1InputStream(new ByteArrayInputStream(oct.getOctets())).readObject();
+	} catch (IOException e)	{
+	    log.error("Trying to obtain policyinfo from certificate failed: "+e.getMessage());
```
```java
LOG.error("Trying to obtain policyinfo from certificate failed", e);
```

msalle replied:
I have left this one, to prevent too many/long stacktraces, which are not really interesting or informative for the issue at hand. We can discuss...

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	// Return oidList
+	return oidList.toArray(new String[0]);
```
IIRC, you can save an allocation here, as you now the size of the oidList:
```java
return oidList.toArray(new String[oidList.size()]);
```

msalle replied:
Not any more apparently: http://stackoverflow.com/questions/4042434/converting-arrayliststring-to-string-in-java/4042464#4042464 see also the bottomline of https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion

andreaceccanti replied:
ok, good to know!

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,104 @@
+
+	// Mandatory argument is missing
+	if (acceptedAttributesValue == null)
+	    throw new ConfigurationException("No "+ACCEPTED_ATTRS_KEY+" specified, nothing to do");
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,104 @@
+
+	// Value is invalid
+	if (acceptedAttrName.length==0)
+	    throw new ConfigurationException("Empty value for "+ACCEPTED_ATTRS_KEY+" specified");
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIPIniConfigurationParser.java:
```diff
@@ -0,0 +1,104 @@
+	// Initialize new acceptedAttribute ArrayList
+	ArrayList<X509ExtractorPIP.AcceptedAttr> acceptedAttrList=new ArrayList<X509ExtractorPIP.AcceptedAttr>();
+	for (int i=0; i<acceptedAttrName.length; i++)	{
+	    if (X509ExtractorPIP.ATTR_X509_ISSUER.equals(acceptedAttrName[i]))
```
braces around blocks (also in other if-else statements)

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+
+
+/**
+ * Subclass for {@link org.glite.authz.pep.pip.provider.PolicyNamesPIP}
```
It is no longer a subclass, comment is wrong

msalle replied:
true, I've now called it helper class, do you agree ?

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+ */
+public class PolicyNamesCache {
+    /** Class logger instance */
+    private final Logger log = LoggerFactory.getLogger(PolicyNamesCache.class);
```
```java
private static final Logger LOG = ...
```

msalle replied:
see above...

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    private final Logger log = LoggerFactory.getLogger(PolicyNamesCache.class);
+
+
+    ////////////////////////////////////////////////////////////////////////
```
Do these comments really provide any info?

msalle replied:
I find it helpful, but we can remove them if you like. I left them in for the time being...

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    private LinkedHashMap<Path,Entry> oldExtInfoEntries = null;
+    
+    /** Directory containing the {@value #FILE_SFX} files */
+    private String trust_dir = null;
```
trust_dir -> trustDir

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    private long initTime = 0;
+
+    /** Number of new entries in the current update */
+    private int nentries_new;
```
camelCase for these members

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    public String[] matchIssuerDN(String issuerDN) {
+	// Protect against empty cache
+	if (infoEntries == null)
+	    return new String[0];
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	    return new String[0];
+
+	// Initialize an empty list of policynames
+	ArrayList<String> policynames = new ArrayList<String>();
```
policynames -> policyNames

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	ArrayList<String> policynames = new ArrayList<String>();
+
+	// Loop over the cached list and look for match
+	for (Entry entry: infoEntries.values())
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+
+	// Loop over the cached list and look for match
+	for (Entry entry: infoEntries.values())
+	    if (entry.subDNs.contains(issuerDN))
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+		policynames.add(entry.name);
+
+	// Convert ArrayList to an array and return
+	return policynames.toArray(new String[0]);
```
```java
policynames.toArray(new String[policyNames.size()]);
```

msalle replied:
See above: apparently nowadays String[0] is the preferred one...

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	}
+
+	// Get current list of info files
+	ArrayList<Path> infofiles;
```
join in the same line, and use generic type:
```java
final List<Path> infoFiles = getInfoFiles(trustDir);
```

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	    try {
+		handleEntry(infofiles.get(i), oldInfoEntries, infoEntries, 0);
+	    } catch (ParseException e)	{
+		log.warn("Syntax error, skipping "+infofiles.get(i));
```
log parse exception:
```java
log.warn("Syntax error, skipping "+infofiles.get(i), e);
```

msalle replied:
done, and using {} instead of +

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+
+	// Reprocess overall subDNs list when something changed
+	if (nentries_new > 0) {
+	    for (Entry entry: infoEntries.values())
```
braces around block

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	}
+	
+	// Log statistics
+	log.debug("Updated list ("+trust_dir+"): "+
```
use placeholders!

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    private ArrayList<Path> getInfoFiles(String trust_dir) throws IOException {
+	// Filter for filtering out .info file. Seems faster than using a glob
+	// in newDirectoryStream()
+	DirectoryStream.Filter<Path> filter=new DirectoryStream.Filter<Path>() {
```
The filter is recreated at every method invocation but this is not really needed AFAIU, i.e. you could have the filter as a static final member of the class.

msalle replied:
good point, done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	};
+
+	// Protect against null trust_dir
+	if (trust_dir==null)
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	try {
+	    stream = Files.newDirectoryStream(Paths.get(trust_dir), filter);
+	} catch(IOException e)	{
+	    throw new IOException("Trust dir has problems: "+e.getMessage());
```
When rethrowing, chain the exception if possible:
```java
throw new IOException("Trust dir has problems: "+e.getMessage(),e);
```

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	// Initialize file array
+	ArrayList<Path> files = new ArrayList<Path>();
+	// Add all entries
+	for (Path entry: stream)
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    {
+	// Protect against recursion
+	if (recursion > MAX_RECURSION)
+	    throw new ParseException("Too many levels of recursion (max. "+MAX_RECURSION+") in "+path, recursion);
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+
+	// Check existence of file
+	if (Files.notExists(path))  {
+	    log.warn("Skipping non-existing "+path.getFileName().toString());
```
use placeholder

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+		entry=parseInfoFile(path, modified);
+		nentries_new++;
+	    } catch (ParseException e)	{
+		log.warn("Syntax error, skipping "+path.getFileName().toString());
```
use placeholder

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	}
+
+	// recursively verify or create its extdeps
+	for (Path extpath: entry.extdeps)
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	String value = getSubjectDNvalue(path);
+
+	// Did we find the KEY?
+	if (value==null || value.isEmpty())
```
braces around blocks

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	try {
+	    reader=Files.newBufferedReader(path, Charset.defaultCharset());
+	} catch (IOException e)	{
+	    throw new IOException("Cannot open "+name+": "+e.getMessage());
```
log before throwing  and chain exception

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	} catch (IOException e)	{
+	    // Try to close, this might throw a new IOException. We're throwing
+	    // one in any case.
+	    reader.close();
```
the potential exception here should be swallowed, in order to properly report the error in the next line

msalle replied:
agreed, done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	    // Try to close, this might throw a new IOException. We're throwing
+	    // one in any case.
+	    reader.close();
+	    throw new IOException("Reading from "+name+" failed: "+e.getMessage());
```
log before throwing and chain exception

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	try {
+	    reader.close();
+	} catch (IOException e)	{
+	    throw new IOException("Closing "+name+" failed: "+e.getMessage());
```
Isn't it better to swallow this? What's the added value of throwing it up?

msalle replied:
agreed, done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+     * @throws ParseException on syntax errors in the value
+     * @throws IOException on read errors with one of the info files
+     */
+    private void parseSubjectDNvalue(Entry entry, String value)
```
this should be called parseSubjectDnValue according to camelCase recommendations

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+	throws ParseException, IOException
+    {
+	// Now parse the value part
+	int pos=0, pos2;
```
probaly a regexp should give more reliable/flexible parsing (in one line)

msalle replied:
Shorter, but a lot slower too. regexps are really slow and inefficient in java (not only in java). Even then, it's not very trivial and you get easily quite unreadable regexps. We can discuss this of course.

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/PolicyNamesCache.java:
```diff
@@ -0,0 +1,625 @@
+    ////////////////////////////////////////////////////////////////////////
+
+    /** Internal representation of info file entries */
+    private class Entry	{
```
This probably could be static

msalle replied:
I don't think so: it is also using the trustDir, infoEntries and extInfoEntries from its parent and they should be instance variables, not class variables.

andreaceccanti replied:
ok, ignore my comment then

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+
+
+/**
+ * Subclass for {@link org.glite.authz.pep.pip.provider.PolicyNamesPIP}
```
is no longer a subclass

msalle replied:
agreed, again made into "Helper class" (see above)

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+ */
+public class UpdatingPolicyNamesCache {
+    /** Class logger instance */
+    private final Logger log = LoggerFactory.getLogger(UpdatingPolicyNamesCache.class);
```
static and final bla bla :)

msalle replied:
see above ...

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+
+    /** Default time interval (in msec) after which info files cache will be
+     * refreshed ({@value}) */
+    public final static long UPDATEINTERVAL = 6*3600*1000;
```
I would use TimeUnit here as well

msalle replied:
Here I agree, since it will be a long.

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+ 
+    /** time interval (in msec) after which info files cache will be
+     * refreshed, default {@link #UPDATEINTERVAL}. */
+    private long update_interval = UPDATEINTERVAL;
```
camelCase

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+    private long update_interval = UPDATEINTERVAL;
+
+    /** Whether a thread is updating the {@link PolicyNamesCache} */
+    private volatile boolean updating = false;
```
updating -> isUpdating

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+     */
+    public String[] findIssuerDN(String dn) throws IOException   {
+	// Update the cache (when needed)
+	updateCache();
```
I would not trigger the update here, I would have a separate thread that runs periodically and just triggers the update

msalle replied:
This is something to be done, I'd prefer indeed if you could fix that.

andreaceccanti replied:
ok!

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+	read.lock();
+	try {
+	    // Protect against empty cache
+	    if (cache == null)  {
```
Is there really a case when the cache can be null?

msalle replied:
Probably not, perhaps this should be an assertion instead, but in any case I don't like NPEs...

andreaceccanti replied:
if you don't like NPEs, you should use an Optional for the cache.
I would drop this, since there's really no chance, if the code is written properly, that the cache is null.

msalle replied:
done

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+	// First check (with appropriate read lock) whether we need to do
+	// anything. If so, create a new cache.
+	read.lock();
+	try {
```
I would drop this in favour of a separate thread that just takes the write lock according to a specified period

msalle replied:
I agree with the separate thread, but the write lock would also block use of the old cache which is still valid... It's not really an issue, since updating the set won't take that long, but still...

andreaceccanti replied:
That wouldn't be a problem

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+	    // out-of-date (i.e. hasn't been updated in the mean time by another
+	    // thread
+	    if (newCache==null)
+		// This probably never happens: exception will have been thrown
```
We should drop the code if it never happens

msalle replied:
It probably never happens, but as C programmer I really don't like NPEs.

andreaceccanti commented on src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java:
```diff
@@ -0,0 +1,210 @@
+		// This probably never happens: exception will have been thrown
+		log.warn("New cache is null, continuing to use old one");
+	    else if (cache.getLifeTime() < update_interval)
+		log.info("Other thread appears to have already updated cache");
```
this can also go away if you use the background updater thread

andreaceccanti commented on src/test/java/org/glite/authz/pep/pip/provider/PolicyNamesPIPTest.java:
```diff
@@ -0,0 +1,197 @@
+		String message= e.getMessage();
+		log.debug("EXPECTED: " + message);
+		log.debug("GOT ("+message.indexOf(element)+"): " + element);
+		assertTrue("PIPProcessingException message does not contain: " + element, message.indexOf(element) >= 0);
```
instead of these general assertTrue/assertFalse, which produce ugly output difficoult to understand, people use the hamcrest matchers library and the assertThat method, which makes the test and output more readable

msalle replied:
Could I ask you to rewrite those? I've mostly copied from how things are currently done in the other PIPs. I agree fully that it's ugly, but then we should change it in all.

andreaceccanti replied:
I agree we should have consistency, but my objective here is to ensure that new code that enters to the codebase satisfies certain requirements. So please try the migration to hamcrest, and if you have questions/doubts I will be happy to help

msalle replied:
I've replaced it in a way similar to https://github.com/italiangrid/voms-api-java/blob/master/src/test/java/org/italiangrid/voms/test/TestVOMSESLineParser.java#L70-L73
so using assertThat with a CoreMatcher. I had to update the junit dependency to 4.11 to get it working.
Is this what you had in mind?

andreaceccanti reviewed (changes requested):
Hi Mischa,
I've reviewed the code more deeply (see comments).

In general things look good, but there are things that should be fixed to align it better to Java recommendations for code and other stuff.

The main comment about the updatingCache is about delegating the update responsibility to a background thread. If you don't want to implement that I can do it.

Thanks!

msalle committed:
Fix comments on argus-pep-server PR#18

Fixing comments from https://github.com/argus-authz/argus-pep-server/pull/18
- camelCase
- use slf4j placeholders
- use String.format()
- use {} in all blocks
- rethrow exceptions (in certain cases)
- log some exceptions before throwing them.
- use TimeUnit to convert hours to millisecs
- rename some vars for clarity

msalle committed:
Missed one block

Fixed one more block missing braces

msalle committed:
Make Logger static (and final)

Make Logger for each PIP private static final.

msalle committed:
Also here make Logger static

msalle committed:
Use hamcrest for asserting exceptions

Use assertThat and a hamcrest matcher instead of a not so readable assertTrue
for testing the message of a exception. assertThat needs a slightly newer junit.

msalle committed:
Remove meta policy info files

The two meta policy info files policy-aspen-birch-cedar-dogwood.info and
policy-aspen-birch-cedar.info will be packaged in an external package.

msalle committed:
Drop test on null cache
